File: app/src/main/java/de/logerbyte/moneyminder/data/db/expense/ExpenseRepo.java
package de.logerbyte.moneyminder.data.db.expense;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import de.logerbyte.moneyminder.data.db.ExpenseDatabase;
import io.reactivex.Observable;

/**
 * Created by logerom on 08.08.18.
 */

@Singleton
public class ExpenseRepo implements ExpenseAPI {

    private final ExpenseDatabase mExpenseDatabase;

    @Inject
    public ExpenseRepo(ExpenseDatabase expenseDatabase) {
        this.mExpenseDatabase = expenseDatabase;
    }

    @Override
    public Observable<List<Expense>> getAllExpense() {
        return Observable.fromCallable(() -> mExpenseDatabase.expenseDao().selectAll());
    }

    @Override
    public Observable<Boolean> insertCashItemIntoDB(Expense expense) {
        return Observable.fromCallable(() -> {
            mExpenseDatabase.expenseDao().insert(expense);
            return true;
        });
    }

    @Override
    public Observable<Boolean> deleteCashItem(long id) {
        return Observable.fromCallable(() -> {
            mExpenseDatabase.expenseDao().delete(id);
            return true;
        });
    }

    @Override
    public Observable<Boolean> updateCashItem(Expense expense) {
        return Observable.fromCallable(() -> {
            mExpenseDatabase.expenseDao().update(expense);
            return true;
        });
    }

    @Override
    public Observable<List<String>> getCategories() {
        return Observable.fromCallable(() -> mExpenseDatabase.expenseDao().selectDistinctCategory());
    }
}

File: app/src/main/java/de/logerbyte/moneyminder/framework/database/ExpenseRepo.kt
package de.logerbyte.moneyminder.framework.database

import androidx.lifecycle.LiveData
import de.logerbyte.moneyminder.entities.data.ExpenseAPI
import io.reactivex.Observable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by logerom on 08.08.18.
 */
@Singleton
class ExpenseRepo @Inject constructor(private val mExpenseDatabase: ExpenseDatabase) :
    ExpenseAPI {

    override fun getAllExpense(): LiveData<List<Expense>> {
        return mExpenseDatabase.expenseDao().selectAll()
    }

    override fun insertCashItemIntoDB(expense: Expense): Observable<Boolean> {
        return Observable.fromCallable {
            mExpenseDatabase.expenseDao().insert(expense)
            true
        }
    }

    override fun deleteCashItem(id: Long): Observable<Boolean> {
        return Observable.fromCallable {
            mExpenseDatabase.expenseDao().delete(id)
            true
        }
    }

    override fun updateCashItem(expense: Expense): Observable<Boolean> {
        return Observable.fromCallable {
            mExpenseDatabase.expenseDao().update(expense)
            true
        }
    }

    override fun getCategories(): Observable<List<String>> {
        return Observable.fromCallable { mExpenseDatabase.expenseDao().selectDistinctCategory() }
    }

    override fun expensesWithCategories(checkedCategories: Set<String>): Observable<MutableList<Expense>>? {
        return Observable.fromCallable { mExpenseDatabase.expenseDao().expensesWithCategories(checkedCategories.toTypedArray()) }
    }
}